### PR TITLE
i#3544 RV64: Support mangling exclusive loads using tp register

### PR DESCRIPTION
**Pull request by ziyao233**
Similar to the way we mangle exclusive loads using the stolen register, spill out a register and load the guest value of tp. One more argument is added to pick_scratch_reg() for cases where two scratch registers have been occupied.

Tests will be included after all corner cases have been covered. For now, a hand-crafted assembly sequence like

```
_start:
        lla             a0,            slot
        mv              tp,             a0
        lr.w.aqrl       s11,            (tp)
        mv              a0,             s11
        li              a7,             93
        ecall
        .align          4
slot:   .long           55
```

could be used as a test by checking return value. Manually checking generated code confirms scratch registers are correctly saved.

Issue: #3544

### CODE REVIEW

**Review by ksco — APPROVED** [review]
LGTM, thanks!

File: core/arch/riscv64/mangle.c [hunk 1]
```diff
@@ -955,9 +966,13 @@ mangle_exclusive_load(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
     ASSERT(opnd_is_reg(dst) && opnd_is_base_disp(src0));
     if (opnd_get_reg(dst) == dr_reg_stolen) {
         opnd_replace_reg(&dst, dr_reg_stolen, scratch_reg2);
+    } else if (opnd_get_reg(dst) == DR_REG_TP) {
+        opnd_replace_reg(&dst, DR_REG_TP, scratch_reg3);
```
Comment (derekbruening) [inline core/arch/riscv64/mangle.c:970]: Could we use instr_replace_reg_resize() which will cover all operands for us?
Then we have just two calls to that, one for dr_reg_stolen and one for DR_REG_TP?

File: core/arch/riscv64/mangle.c [hunk 2]
```diff
@@ -1021,8 +1044,8 @@ mangle_exclusive_store(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
     dst1 = instr_get_dst(instr, 1);
     ASSERT(opnd_is_base_disp(dst0));
     opsz = instr_get_opcode(instr) == OP_sc_d ? OPSZ_8 : OPSZ_4;
-    scratch_reg1 = pick_scratch_reg(dcontext, instr, DR_REG_NULL, &slot1);
-    scratch_reg2 = pick_scratch_reg(dcontext, instr, scratch_reg1, &slot2);
+    scratch_reg1 = pick_scratch_reg(dcontext, instr, DR_REG_NULL, DR_REG_NULL, &slot1);
+    scratch_reg2 = pick_scratch_reg(dcontext, instr, scratch_reg1, DR_REG_NULL, &slot2);
 
     /* Spill scratch registers. */
     PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg1, slot1));
```
Comment (derekbruening) [inline core/arch/riscv64/mangle.c:1051]: Explain how this was tested in the PR description.  I see it says an automated test is forthcoming but there must have been manual testing of this change: explain how that was done, to give confidence that this code works.

**Review by derekbruening — APPROVED** [review]

File: core/arch/riscv64/mangle.c (hunk 1 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c:970]: Thanks for your hint! But currently riscv implementation of `opnd_replace_reg_resize()` is a stub[1], thus `instr_replace_reg_resize()` doesn't work. I'm willing to defer this change until the stub is replaced with a real implementation.

[1]: https://github.com/DynamoRIO/dynamorio/blob/4e6e71d337a4dcf46d7361d03c0ba0b354759a39/core/ir/opnd_shared.c#L1586-L1600

File: core/arch/riscv64/mangle.c (hunk 2 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c:1051]: Okay, the description has been updated with a short description of testing.

File: core/arch/riscv64/mangle.c (hunk 1 above)
Comment (derekbruening) [reply core/arch/riscv64/mangle.c:970]: Or add a XXX comment here to replace when that is available for simpler code and go ahead and merge this PR.

**Review by derekbruening — APPROVED** [review]

File: core/arch/riscv64/mangle.c (hunk 1 above)
Comment (ziyao233) [reply core/arch/riscv64/mangle.c:972]: > Or add a XXX comment here to replace when that is available for simpler code and go ahead and merge this PR.

Done
